File: Tests/acc_async_test_all.c
#include "acc_testsuite.h"
#ifndef T1
//T1:async,runtime,construct-independent,V:2.0-2.7
int test1(){
    int err = 0;
    real_t **a = (real_t **)malloc(10 * sizeof(real_t*));
    real_t **b = (real_t **)malloc(10 * sizeof(real_t*));
    real_t **c = (real_t **)malloc(10 * sizeof(real_t*));
    real_t **d = (real_t **)malloc(10 * sizeof(real_t*));
    real_t **e = (real_t **)malloc(10 * sizeof(real_t*));

    for (int x = 0; x < 10; ++x){
        a[x] = (real_t *)malloc(n * sizeof(real_t));
        b[x] = (real_t *)malloc(n * sizeof(real_t));
        c[x] = (real_t *)malloc(n * sizeof(real_t));
        d[x] = (real_t *)malloc(n * sizeof(real_t));
        e[x] = (real_t *)malloc(n * sizeof(real_t));
        for (int y = 0; y < n; ++y){
            a[x][y] = rand() / (real_t)(RAND_MAX / 10);
            b[x][y] = rand() / (real_t)(RAND_MAX / 10);
            c[x][y] = 0;
            d[x][y] = rand() / (real_t)(RAND_MAX / 10);
            e[x][y] = 0;
        }
    }

    #pragma acc enter data create(a[0:10][0:n], b[0:10][0:n], c[0:10][0:n], d[0:10][0:n], e[0:10][0:n])
    for (int x = 0; x < 10; ++x){
        #pragma acc update device(a[x:1][0:n], b[x:1][0:n], d[x:1][0:n]) async(x)
        #pragma acc parallel present(a[x:1][0:n], b[x:1][0:n], c[x:1][0:n]) async(x)
        {
            #pragma acc loop
            for (int y = 0; y < n; ++y){
                c[x][y] = a[x][y] + b[x][y];
            }
        }
        #pragma acc parallel present(c[x:1][0:n], d[x:1][0:n], e[x:1][0:n]) async(x)
        {
            #pragma acc loop
            for (int y = 0; y < n; ++y){
                e[x][y] = c[x][y] + d[x][y];
            }
        }
        #pragma acc update host(e[x:1][0:n]) async(x)
    }
   
    while(!acc_async_test_all());
    for (int x = 0; x < 10; ++x){
        for (int y = 0; y < n; ++y){
            if (fabs(e[x][y] - (a[x][y] + b[x][y] + d[x][y])) > PRECISION){
                err += 1;
            }
        }
    }
    #pragma acc exit data delete(a[0:10][0:n], b[0:10][0:n], c[0:10][0:n], d[0:10][0:n], e[0:10][0:n])
    for (int x = 0; x < 10; ++x){
        free(a[x]);
        free(b[x]);
        free(c[x]);
        free(d[x]);
        free(e[x]);
    }
    free(a);
    free(b);
    free(c);
    free(d);
    free(e);
    
    return err;
}
#endif

#ifndef T2
//T2:async,runtime,construct-independent,V:1.0-2.7
int test2(){
    int err = 0;
    real_t *a = (real_t *)malloc(10 * n * sizeof(real_t));
    real_t *b = (real_t *)malloc(10 * n * sizeof(real_t));
    real_t *c = (real_t *)malloc(10 * n * sizeof(real_t));
    real_t *d = (real_t *)malloc(10 * n * sizeof(real_t));
    real_t *e = (real_t *)malloc(10 * n * sizeof(real_t));

    for (int x = 0; x < 10*n; ++x){
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = rand() / (real_t)(RAND_MAX / 10);
        c[x] = 0;
        d[x] = rand() / (real_t)(RAND_MAX / 10);
        e[x] = 0;
    }

    #pragma acc data copyin(a[0:10*n], b[0:10*n], d[0:10*n]) copyout(c[0:10*n], e[0:10*n])
    {
        for (int x = 0; x < 10; ++x){
            #pragma acc parallel present(a[0:10*n], b[0:10*n], c[0:10*n]) async(x)
            {
                #pragma acc loop
                for (int y = 0; y < n; ++y){
                    c[x * n + y] = a[x * n + y] + b[x * n + y];
                }
            }
            #pragma acc parallel present(c[0:10*n], d[0:10*n], e[0:10*n]) async(x)
            {
                #pragma acc loop
                for (int y = 0; y < n; ++y){
                    e[x * n + y] = c[x * n + y] + d[x * n + y];
                }
            }
        }
        while(!acc_async_test_all());
    }

    for (int x = 0; x < 10*n; ++x){
        if (fabs(e[x] - (a[x] + b[x] + d[x])) > PRECISION){
            err += 1;
        }
    }
    
    free(a);
    free(b);
    free(c);
    free(d);
    free(e);
    
    return err;
}
#endif

#ifndef T3
//T3:async,runtime,construct-independent,V:2.5-2.7
int test3(){
    int err = 0;
    real_t **a = (real_t **)malloc(10 * sizeof(real_t *));
    real_t **b = (real_t **)malloc(10 * sizeof(real_t *));
    real_t **c = (real_t **)malloc(10 * sizeof(real_t *));
    real_t **d = (real_t **)malloc(10 * sizeof(real_t *));
    real_t **e = (real_t **)malloc(10 * sizeof(real_t *));

    for (int x = 0; x < 10; ++x){
        a[x] = (real_t *)malloc(n * sizeof(real_t));
        b[x] = (real_t *)malloc(n * sizeof(real_t));
        c[x] = (real_t *)malloc(n * sizeof(real_t));
        d[x] = (real_t *)malloc(n * sizeof(real_t));
        e[x] = (real_t *)malloc(n * sizeof(real_t));
        for (int y = 0; y < n; ++y){
            a[x][y] = rand() / (real_t)(RAND_MAX / 10);
            b[x][y] = rand() / (real_t)(RAND_MAX / 10);
            c[x][y] = 0;
            d[x][y] = rand() / (real_t)(RAND_MAX / 10);
            e[x][y] = 0;
        }
    }

    #pragma acc data copyin(a[0:10][0:n], b[0:10][0:n], d[0:10][0:n]) copyout(c[0:10][0:n], e[0:10][0:n])
    {
        for (int x = 0; x < 10; ++x){
            acc_set_default_async(x);
            #pragma acc parallel present(a[0:10][0:n], b[0:10][0:n], c[0:10][0:n]) async
            {
                #pragma acc loop
                for (int y = 0; y < n; ++y){
                    c[x][y] = a[x][y] + b[x][y];
                }
            }
            #pragma acc parallel present(c[0:10][0:n], d[0:10][0:n], e[0:10][0:n]) async
            {
                #pragma acc loop
                for (int y = 0; y < n; ++y){
                    e[x][y] = c[x][y] + d[x][y];
                }
            }
        }
        while (!acc_async_test_all());
    }

    int count = 0;
    int total = 0;
    for (int x = 0; x < 10; ++x) {
        for (int y = 0; y < n; ++y) {
            if (fabs(e[x][y]) < PRECISION) {
                count += 1;
            }
            total += 1;
        }
    }
    printf("%d out of %d\n", count, total);

    for (int x = 0; x < 10; ++x){
        for (int y = 0; y < n; ++y){
            if (fabs(e[x][y] - (a[x][y] + b[x][y] + d[x][y])) > PRECISION){
                err += 1;
            }
        }
    }
    
    for (int x = 0; x < 10; ++x){
        free(a[x]);
        free(b[x]);
        free(c[x]);
        free(d[x]);
        free(e[x]);
    }
    free(a);
    free(b);
    free(c);
    free(d);
    free(e);
    
    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
#ifndef T2
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test2();
    }
    if (failed != 0){
        failcode = failcode + (1 << 1);
    }
#endif
#ifndef T3
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test3();
    }
    if (failed != 0){
        failcode = failcode + (1 << 2);
    }
#endif
    return failcode;
}

File: Tests/acc_async_test_all.f90
      INTEGER FUNCTION test()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.fh"
        INTEGER :: x, y !Iterators
        REAL(8),DIMENSION(10, LOOPCOUNT):: a, b, c, d, e !Data
        REAL(8) :: RAND
        INTEGER :: errors = 0

        !Initilization
        CALL RANDOM_SEED()
        CALL RANDOM_NUMBER(a)
        CALL RANDOM_NUMBER(b)
        CALL RANDOM_NUMBER(d)
        c = 0
        e = 0

        !$acc enter data create(a(1:10,1:LOOPCOUNT), b(1:10,1:LOOPCOUNT), c(1:10,1:LOOPCOUNT), d(1:10,1:LOOPCOUNT), e(1:10,1:LOOPCOUNT))
        DO x = 1, 10
          !$acc update device(a(x:x,1:LOOPCOUNT), b(x:x,1:LOOPCOUNT), d(x:x,1:LOOPCOUNT)) async(x)
          !$acc parallel present(a(x:x,1:LOOPCOUNT), b(x:x,1:LOOPCOUNT), c(x:x,1:LOOPCOUNT)) async(x)
            !$acc loop
            DO y = 1, LOOPCOUNT
              c(x, y) = a(x, y) + b(x, y)
            END DO
          !$acc end parallel
          !$acc parallel present(c(x:x,1:LOOPCOUNT), d(x:x,1:LOOPCOUNT),e(x:x,1:LOOPCOUNT)) async(x)
            !$acc loop
            DO y = 1, LOOPCOUNT
              e(x, y) = c(x, y) + d(x, y)
            END DO
          !$acc end parallel
          !$acc update host(e(x:x,1:LOOPCOUNT)) async(x)
        END DO

        DO WHILE (acc_async_test_all() .eqv. .FALSE.)

        END DO


        DO x = 1, 10
          DO y = 1, LOOPCOUNT
            IF (abs(e(x, y) - (a(x, y) + b(x, y) + d(x, y))) .GT. PRECISION) THEN
              errors = errors + 1
            END IF
          END DO
        END DO
        test = errors
      END


      PROGRAM test_kernels_async_main
      IMPLICIT NONE
      INTEGER :: failed, success !Number of failed/succeeded tests
      INTEGER :: num_tests,crosschecked, crossfailed, j
      INTEGER :: temp,temp1
      INCLUDE "acc_testsuite.fh"
      INTEGER test


      CHARACTER*50:: logfilename !Pointer to logfile
      INTEGER :: result

      num_tests = 0
      crosschecked = 0
      crossfailed = 0
      result = 1
      failed = 0

      !Open a new logfile or overwrite the existing one.
      logfilename = "test.log"
!      WRITE (*,*) "Enter logFilename:"
!      READ  (*,*) logfilename

      OPEN (1, FILE = logfilename)

      WRITE (*,*) "######## OpenACC Validation Suite V 1.0a ######"
      WRITE (*,*) "## Repetitions:", N
      WRITE (*,*) "## Loop Count :", LOOPCOUNT
      WRITE (*,*) "##############################################"
      WRITE (*,*)

      WRITE (*,*) "--------------------------------------------------"
      !WRITE (*,*) "Testing acc_kernels_async"
      WRITE (*,*) "Testing test_kernels_async"
      WRITE (*,*) "--------------------------------------------------"

      crossfailed=0
      result=1
      WRITE (1,*) "--------------------------------------------------"
      !WRITE (1,*) "Testing acc_kernels_async"
      WRITE (1,*) "Testing test_kernels_async"
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*)
      WRITE (1,*) "testname: test_kernels_async"
      WRITE (1,*) "(Crosstests should fail)"
      WRITE (1,*)

      DO j = 1, N
        temp =  test()
        IF (temp .EQ. 0) THEN
          WRITE (1,*)  j, ". test successfull."
          success = success + 1
        ELSE
          WRITE (1,*) "Error: ",j, ". test failed."
          failed = failed + 1
        ENDIF
      END DO


      IF (failed .EQ. 0) THEN
        WRITE (1,*) "Directive worked without errors."
        WRITE (*,*) "Directive worked without errors."
        result = 0
        WRITE (*,*) "Result:",result
      ELSE
        WRITE (1,*) "Directive failed the test ", failed, " times."
        WRITE (*,*) "Directive failed the test ", failed, " times."
        result = failed * 100 / N
        WRITE (*,*) "Result:",result
      ENDIF
      CALL EXIT (result)
      END PROGRAM
